pkg/usagetracker/usagetrackerclient/client.go

```diff
@@ -281,3 +371,113 @@ func (c *UsageTrackerClient) sortZones(zones []string) []string {
+		if err != nil {
+			return nil, errors.Errorf("usage-tracker instance %s (%s)", instance.Id, instance.Addr)
+		}
+
+		trackerClient := poolClient.(usagetrackerpb.UsageTrackerClient)
+		resp, err := trackerClient.GetUsersCloseToLimit(ctx, &usagetrackerpb.GetUsersCloseToLimitRequest{Partition: partitionID})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get users close to limit from partition %d", partitionID)
+		}
+
+		c.usersCloseToLimitsMtx.Lock()
+		first := !c.usersCloseToLimitLoaded
+		c.usersCloseToLimit = slices.Clone(resp.UserIds)
+		c.usersCloseToLimitLoaded = true
+		c.usersCloseToLimitsMtx.Unlock()
+
+		// Update metrics.
+		c.usersCloseToLimitCount.Set(float64(len(resp.UserIds)))
+		c.usersCloseToLimitLastUpdateSeconds.Set(float64(time.Now().Unix()))
+
+		lvl := level.Debug
+		if first {
+			lvl = level.Info
+		}
+		lvl(c.logger).Log("msg", "updated users close to limit cache", "partition", resp.Partition, "user_count", len(resp.UserIds))
+		return nil, nil
+	}, func([]string) {})
+
+	if err != nil {
+		c.usersCloseToLimitUpdateFailures.Inc()
+		level.Error(c.logger).Log("msg", "failed to get users close to limit from usage-tracker", "err", err)
+		return false
+	}
+
+	return true
+}
+
+func (c *UsageTrackerClient) selectRandomPartition() (int32, ring.ReplicationSet, bool) {
+	partitions := c.partitionRing.PartitionRing().ActivePartitionIDs()
+	if len(partitions) == 0 {
+		level.Error(c.logger).Log("msg", "no partitions available in ring for users close to limit poll")
+		return 0, ring.ReplicationSet{}, false
+	}
+	partitionID := partitions[rand.IntN(len(partitions))]
+	set, err := c.partitionRing.GetReplicationSetForPartitionAndOperation(partitionID, TrackSeriesOp)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to get replication set for partition", "partition", partitionID, "err", err)
+		return 0, ring.ReplicationSet{}, false
+	}
+
+	return partitionID, set, true
+}
+
+// CanTrackAsync returns true if the user can be tracked asynchronously.
+// A user can be tracked async if:
+// 1. The user is NOT in the cache of users close to their limit, AND
+// 2. The user's series limit is >= MinSeriesLimitForAsyncTracking (if configured)
+func (c *UsageTrackerClient) CanTrackAsync(userID string) bool {
+	// Check if user's limit is below the minimum threshold for async tracking.
+	if c.cfg.MinSeriesLimitForAsyncTracking > 0 {
+		userLimit := c.limits.MaxActiveSeriesPerUser(userID)
+		if userLimit > 0 && userLimit < c.cfg.MinSeriesLimitForAsyncTracking {
+			// User's limit is too low, must track synchronously.
+			return false
+		}
+	}
+
+	// Check if user is close to their limit.
+	c.usersCloseToLimitsMtx.RLock()
+	defer c.usersCloseToLimitsMtx.RUnlock()
+	if !c.usersCloseToLimitLoaded {
+		// Can't do async if we don't really know.
+		return false
+	}
+
+	// Can track async if it's not in the list of users close to their limit.
+	_, found := slices.BinarySearch(c.usersCloseToLimit, userID)
```

Comment:
Is `usersCloseToLimit` guaranteed to be sorted? Are we sure we won't break this guarantee by accident? Maybe naming should reflect this?

Apparently it is, but I had to go all the way here to make sure:

https://github.com/grafana/mimir/blob/dec0aedd8e9a8c71687bc2757659aa2f8065411f/pkg/usagetracker/tracker_store.go#L269

Reply:
I made it explicit through naming at all levels: https://github.com/grafana/mimir/pull/13427/commits/6f501d423904fc083b2f9cec524d1fccc09f23e5